firmware/f7/Cam/Src/F7Cam_Driver.c

```diff
@@ -0,0 +1,737 @@
+#include "F7Cam_Driver.h"
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "camera_I2C.h"
+#include "events.h"
+#include "stm32f7xx_hal.h"
+#include "stm32f7xx_hal_dcmi.h"
+
+#define OV5640
+
+char cam_fb[CAM_FB_SIZE];// __attribute__ ((section (".sdram"), aligned (4)));
+// uint8_t jpeg_search_buffer[CAM_FB_SIZE];
+uint8_t frameCounter = 0;
+uint8_t buffer_offset;
+uint8_t packetCounter = 0;
+
+void DCMI_DMA_TRANSFER_HALF_COMPLETE(DMA_HandleTypeDef *hdma);
+void DCMI_DMA_TRANSFER_COMPLETE(DMA_HandleTypeDef *hdma);
+static void DCMI_DMAError(DMA_HandleTypeDef *hdma);
+
+bool LoadedFlag = false;
+uint8_t loadingCounter = 0;
+
+DCMI_HandleTypeDef  hDcmiHandler;
+CAMERA_DrvTypeDef   *camera_drv;
+
+uint32_t JPEG_counter = 0;
+uint32_t printFrameCounter = 0;
+uint32_t current_JPEG_size = 0;
+uint8_t bufferblockcounter;
+
+uint16_t JPEG_Size = 0;
+uint16_t JPEG_FIFO_SIZE_BUFFER[20];
+uint16_t JPEG_HEX_SIZE_BUFFER[20];
+uint16_t JPEG_PACKET_COUNT_BUFFER[20];
+
+#ifdef OV5640
+uint8_t CAM_Init(uint8_t format, uint16_t x_res, uint16_t y_res, uint8_t FPS, uint16_t FB_size, uint16_t FIFO_width, uint8_t jpeg_comp_ratio)
+{
+  DCMI_HandleTypeDef *phdcmi;
+  uint8_t status = CAMERA_ERROR;
+
+  /* Get the DCMI handle structure */
+  phdcmi = &hDcmiHandler;
+
+  /*** Configures the DCMI to interface with the camera module ***/
+  /* DCMI configuration */
+  phdcmi->Init.CaptureRate      = DCMI_CR_ALL_FRAME;
+  phdcmi->Init.HSPolarity       = DCMI_HSPOLARITY_HIGH;
+  phdcmi->Init.SynchroMode      = DCMI_SYNCHRO_HARDWARE;
+  phdcmi->Init.VSPolarity       = DCMI_VSPOLARITY_HIGH;
+  phdcmi->Init.ExtendedDataMode = DCMI_EXTEND_DATA_8B;
+  phdcmi->Init.PCKPolarity      = DCMI_PCKPOLARITY_RISING;
+   if (format == FMT_JPEG) {
+	   phdcmi->Init.JPEGMode 		  = DCMI_JPEG_ENABLE;
+   }
+  phdcmi->Instance              = DCMI;
+  status = CAMERA_ERROR;
+	BSP_CAMERA_PwrDown();
+  BSP_CAMERA_PwrUp();
+	HAL_Delay(1000);
+  /* Read ID of Camera module via I2C */
+  if(ov5640_ReadID(CAMERA_I2C_ADDRESS_OV5640) == OV5640_ID)
+  {
+    /* Initialize the camera driver structure */
+    //camera_drv = &ov5640_drv;
+    //CameraHwAddress = CAMERA_I2C_ADDRESS_OV5640;
+  
+	    switch (format)
+	        {
+	    		case FMT_JPEG:
+	    		{
+	    			ov5640_Init_JPEG(x_res, y_res, FIFO_SIZE, PACKET_COUNT, jpeg_comp_ratio);
+            if (FPS != 0) {
+              OV5640_SetPCLK(FPS);

```

Comment:
`OV5640_SetPCLK` takes the `OV5640_PCLK_*` macro, not the desired FPS. I think we maybe need a couple of intermediate functions here:
1. One to go from the desired FPS to the desired PCLK frequency. Is this dependent on the image settings at all? Even if not I think having the function would improve readability.
2. One to go from the desired PCLK to frequency to the required PLL multiplier and divider

I think 2. is quite nice to write about in the report as there is some maths involved and possibly nice equations to put in or maybe we have to use numerical methods. Also for the report, as PCLK is exposed on one of the camera module's GPIOs it can be measured and graphed.

I think with those two functions implemented we'd get much more granular control over the FPS also.